book-content/chapters/04-essential-types-and-annotations.md

```diff
@@ -0,0 +1,2366 @@
+
+One way to have TypeScript know what type we want the `Set` to hold would be to pass in some initial values:
+
+```ts twoslash
+const formats = new Set(["CD", "DVD"]);
+//    ^?
+```
+
+In this case, since we specified two strings when creating the `Set`, TypeScript knows that `formats` is a `Set` of strings.
+
+But it's not always the case that we know exactly what values we want to pass to a `Set` when we create it. We might want to create an empty `Set` that we know will hold strings later on.
+
+For this, we can pass a type to `Set` using the angle brackets syntax:
+
+```ts
+const formats = new Set<string>();
+```
+
+Now, `formats` understands that it's a set of strings, and adding anything other than a string will fail:
+
+```ts twoslash
+// @errors: 2345
+const formats = new Set<string>();
+
+// ---cut---
+formats.add("Digital");
+
+formats.add(8);
+```
+
+This is a really important thing to understand in TypeScript. You can pass types, as well as values, to functions.
+
+### Not All Functions Can Receive Types
+
+Most functions in TypeScript _can't_ receive types.
+
+For example, let's look at `document.getElementById` that comes in from the DOM typings.
+
+A common example where you might want to pass a type is when calling `document.getElementById`. Here we're trying to get an audio element:
+
+```ts twoslash
+const audioElement = document.getElementById("player");
+```
+
+We know that `audioElement` is going to be a `HTMLAudioElement`, so it seems like we should be able to pass it to `document.getElementById`:
+
+```ts twoslash
+// @errors: 2558
+const audioElement = document.getElementById<HTMLAudioElement>("player");
+```
+
+But unfortunately, we can't. We get an error saying that `.getElementById` expects zero type arguments.
+
+We can see whether a function can receive type arguments by hovering over it. Let's try hovering `.getElementById`:
+
+```ts
+// hovering over .getElementById shows:
+(method) Document.getElementById(elementId: string): HTMLElement | null
+```
+
+Notice that `.getElementById` contains no angle brackets (`<>`) in its hover, which is why we can't pass a type to it.
+
+Let's contrast it with a function that _can_ receive type arguments, like `document.querySelector`:
+
+```ts
+const audioElement = document.querySelector("#player");
+
+// hovering over .querySelector shows:
+(method) ParentNode.querySelector<Element>(selectors: string): Element | null
+```
+
+This type definition shows us that `.querySelector` has some angle brackets before the parentheses. Inside of the brackets is the default value inside them - in this case, `Element`.
+
+So, to fix our code above we could replace `.getElementById` with `.querySelector` and use the `#player` selector to find the audio element:
+
+```ts
+const audioElement = document.querySelector<HTMLAudioElement>("#player");
+```
+
+And everything works.
```

Comment:
it could be helpful to mention here that this doesn't truly provide type-safety, we just convince TS to trust us on this one